### PR TITLE
upgrade openssl from 1.0.2k to 1.1.1m and related stuff

### PR DESCRIPTION
**Pull request by rzuo**
Backgrounds:
1. Refer to PR 2733(https://github.com/StarRocks/starrocks/pull/2733) first.
2. When compiling mariadb-connector-c 3.1.14 on ARM platform, it failed with openssl "relocation of symbol OPENSSL_armcap_P" error.
3. Openssl has fixed the relocation issue by PR 13056(https://github.com/openssl/openssl/pull/13056) and merged to release of openssl after version 1.1.1i.
4. PR 2733 fixed the above problem by bypassing the asm part of SHA.
5. To fix this issue completely, decided to upgrade openssl to 1.1.1m.

This fix includes following modifications:
1. upgrade openssl             from 1.0.2k to 1.1.1m
   * simplify build_openssl() function in thirdparty/build-thirdparty.sh
   * modify AesUtil::entrypt() and AesUtil::decrypt() according to the API changes of openssl from 1.0.x to 1.1.x
2. upgrade curl                from 7.54.1 to 7.79.0
3. upgrade mariadb-connector-c from 3.1.14 to 3.2.5
   * modify build_mariadb() function in thirdparty/build-thirdparty.sh for static link
   * add a patch mariadb-connector-c-3.2.5-for-starrocks-static-link.patch for static link
     * refer to https://mariadb.com/kb/en/configuration-settings-for-building-connectorc/
4. add platform check for be/src/formats module for compatible with ARM platform
5. verified on:
   * CentOS 7.9.2009(Core) / i5-7200U
   * Kylin Linux Advanced Server release V10(Tercel) / FT-2000+/64

TODOs
1. Need a complete test, especially for test cases related TLS/SSL communication.
2. The performance after this thirdparty libraries upgrade is currently unknown, need more measurement by performance test.

Potential risks
1. All newly imported thirdparty libraries need be compatible with openssl 1.1.x APIs.

### CODE REVIEW

File: be/src/util/aes_util.cpp [hunk 1]
```diff
@@ -113,10 +113,10 @@ int AesUtil::encrypt(AesMode mode, const unsigned char* source, uint32_t source_
     if (cipher == nullptr || (EVP_CIPHER_iv_length(cipher) > 0 && !iv)) {
         return AES_BAD_DATA;
     }
-    EVP_CIPHER_CTX_init(&aes_ctx);
+    EVP_CIPHER_CTX_reset(aes_ctx);
```
Comment (rickif) [inline be/src/util/aes_util.cpp:116]: Seems that `EVP_CIPHER_CTX_reset` is no longer needed here.

File: be/src/util/aes_util.cpp [hunk 2]
```diff
@@ -160,10 +160,10 @@ int AesUtil::decrypt(AesMode mode, const unsigned char* encrypt, uint32_t encryp
         return AES_BAD_DATA;
```
Comment (imay) [inline be/src/util/aes_util.cpp:160]: aes_ctx will leak?
Better to create aes_ctx after this if clause

File: thirdparty/download-thirdparty.sh [hunk 3]
```diff
@@ -339,6 +339,7 @@ if [ ! -f $PATCHED_MARK ] && [ $GPERFTOOLS_SOURCE = "gperftools-gperftools-2.7"
     touch $PATCHED_MARK
 fi
 cd -
+echo "Finished patching $GPERFTOOLS_SOURCE"
 
 # patch openssl-1.0.2k
```
Comment (rickif) [inline thirdparty/download-thirdparty.sh:344]: The patch of openssl-1.0.x could be removed.

File: be/src/util/aes_util.cpp (hunk 1 above)
Comment (rzuo) [reply be/src/util/aes_util.cpp:116]: I've checked the document of openssl about EVP_CIPHER_CTX_reset function. And got following explanations:
https://www.openssl.org/docs/man1.1.1/man3/EVP_CIPHER_CTX_reset.html

```
EVP_CIPHER_CTX_reset() clears all information from a cipher context and free up any allocated memory associate with it, except the ctx itself. 
This function should be called anytime ctx is to be reused for another EVP_CipherInit() / EVP_CipherUpdate() / EVP_CipherFinal() series of calls.
```

So, I think we can keep this function call as is.

Comment (rickif) [reply be/src/util/aes_util.cpp:116]: The init/clear is already done in `EVP_CIPHER_CTX_new`. And there's no reusing of `aes_ctx` here.

File: be/src/util/aes_util.cpp (hunk 2 above)
Comment (rzuo) [reply be/src/util/aes_util.cpp:160]: Yes, it's possible of leaking. Will correct it and commit again.

File: thirdparty/download-thirdparty.sh (hunk 3 above)
Comment (rzuo) [reply thirdparty/download-thirdparty.sh:344]: Yes, this can be removed.

Comment (rickif) [reply thirdparty/download-thirdparty.sh]: File `thirdparty/patches/openssl-1.0.2k.patch` could be also removed.

**Review by imay — DISMISSED** [review]
LGTM

**Comment by rickif** [comment]
LGTM!

**Pull request merged**